SimPEG/ObjectiveFunction.py

```diff
@@ -0,0 +1,258 @@
+
+        return checkDerivative(
+            lambda m: [self.deriv(m), self.deriv2(m)], x, num=num,
+            plotIt=plotIt
+        )
+
+    def test(self, x=None, num=4, plotIt=False, **kwargs):
+        deriv = self._test_deriv(x=x, num=num, **kwargs)
+        deriv2 = self._test_deriv2(x=x, num=num, plotIt=False, **kwargs)
+        return (deriv & deriv2)
+
+    def __add__(self, objfct2):
+
+        if isinstance(objfct2, Utils.Zero):
+            return self
+
+        if not isinstance(objfct2, BaseObjectiveFunction):
+            raise Exception(
+                "Cannot add type {} to an objective function. Only "
+                "ObjectiveFunctions can be added together".format(
+                    objfct2.__class__.__name__
+                )
+            )
+
+        if isinstance(self, ComboObjectiveFunction):
+
+            if isinstance(objfct2, ComboObjectiveFunction):
+                objfctlist = self.objfcts + objfct2.objfcts
+                multipliers = self.multipliers + objfct2.multipliers
+
+            elif isinstance(objfct2, ObjectiveFunction):
+                objfctlist = self.objfcts.append(objfct2)
+                multipliers = self.multipliers.append(1)
+
+        else:
+            if isinstance(objfct2, ComboObjectiveFunction):
+                objfctlist = [self] + objfct2.objfcts
+                multipliers = [1] + objfct2.multipliers
+
+            else:
+                objfctlist = [self, objfct2]
+                multipliers = None
+
+        return ComboObjectiveFunction(
+            objfcts=objfctlist, multipliers=multipliers
+        )
+
+    def __radd__(self, objfct2):
+        return self+objfct2
+
+    def __mul__(self, multiplier):
+        return ComboObjectiveFunction([self], [multiplier])
+
+    def __rmul__(self, multiplier):
+        return self * multiplier
+
+    def __div__(self, denominator):
+        return self.__mul__(1./denominator)
+
+
+class ComboObjectiveFunction(BaseObjectiveFunction):
+
+    _multiplier_types = (float, None, Utils.Zero) + integer_types # Directive
+
+    def __init__(self, objfcts=[], multipliers=None, **kwargs):
+
+        self.objfcts = []
+        nP = '*'
+
+        for fct in objfcts:
+            assert isinstance(fct, BaseObjectiveFunction), (
+                "Unrecognized objective function type {} in objfcts. All "
+                "entries in objfcts must inherit from "
+                "ObjectiveFunction".format(fct.__class__.__name__)
+            )
+
+            # ensure all objective functions have the same nP
+            if fct.nP != '*':
+                if nP != '*':
+                    assert nP == fct.nP, (
+                        "Objective Functions must all have the same nP, not "
+                        "{}".format([f.nP for f in objfcts])
+                    )
+                else:
+                    nP = fct.nP
+            self.objfcts.append(fct)
+
+        if multipliers is None:
```

Comment:
When a single objective function is evaluated, we want to be able to see the output prior to taking a product with a multiplier (e.g. in the iteration printers). As a simple example, we want to print `phi_d`, `phi_m` and `beta` independently, which seemed much less transparent to me if beta is attached to `phi_m`: what should `__call__` do? 

Alternatively, it can be useful to print the smallness, and smoothness_x, smoothness_y, ... independently to ensure that alphas are chosen correctly. 

So we do need a way to evaluate the objective function and later take a product with a scalar in a transparent way. My thinking in putting it on the `ComboObjectiveFunction` is that it really only makes sense to use a multiplier if we are combining objective functions - otherwise it really has no impact on the inversion.